exercises/supervised-classification/ex_rnw/ex_logreg_dec_boundaries_enhanced.Rnw

```diff
@@ -0,0 +1,145 @@
+  
+  dens <- expand.grid(x_1 = x, x_2 = x)
+  dens$z <- apply(as.matrix(dens), 1L, compute_softmax, theta = theta)
+  d <- akima::interp(x = dens$x_2, y = dens$x_1, z = dens$z)
+  
+  # define perspective plot is viewed from
+  
+  scene = list(
+    camera = list(eye = eye),
+    xaxis = list(title = "x1"),
+    yaxis = list(title = "x2"),
+    zaxis = list(title = "s(f(x1,x2))"))
+  
+  # define color
+  
+  my_palette = c("cornflowerblue", "blue4")
+  
+  # plot
+  
+  plotly::plot_ly(x = d$x, y = d$y, z = d$z) %>%
+    add_surface(
+      showscale = FALSE,
+      colors = my_palette
+      ) %>%
+    layout(scene = scene)
+  
+}
+
+# saving plotly objects requires orca installation, which might not work on 
+# every device
+# here: open the respective plot in your browser (clicking on "show in new 
+# window" in the viewer pane) and save snapshot via the camera symbol without 
+# zooming
+
+p_1 <- plot_3d(seq_x, theta$theta_1)
+p_2 <- plot_3d(seq_x, theta$theta_2)
+p_3 <- plot_3d(seq_x, theta$theta_3)
+
+# add hyperplanes marking decision boundaries for different thresholds
+# use dirty trick with points, cannot get it to work with actual surface :]
+
+add_vertical_hyperplane <- function(plot, 
+                                    n_points = 100L, 
+                                    theta_1 = 0.3, 
+                                    threshold = 0.5) {
+  
+  y_vert <- seq(-10L, 10L, length.out = n_points)
+  z_vert <- seq(0L, threshold, length.out = n_points)
+  yz_vert <- expand.grid(y_vert, z_vert)
+  
+  plot %>%
+  add_trace(
+    inherit = FALSE,
+    x = rep((- 1 / theta_1 * log(1 / threshold - 1)), n_points^2),
+    y = yz_vert[, 1],
+    z = yz_vert[, 2],
+    marker = list(
+      type = "marker",
+      mode = "scatter3d",
+      color = "gray",
+      size = 0.8),
+    showlegend = FALSE) %>%
+    add_trace(
+      inherit = FALSE,
+      x = rep((- 1 / theta_1 * log(1 / threshold - 1)), n_points^2),
+      y = yz_vert[, 1],
+      z = 0L,
+      marker = list(
+        type = "marker",
+        mode = "scatter3d",
+        color = "black",
+        size = 2L),
+      showlegend = FALSE)
+  
+}
+
+p_4 <- add_vertical_hyperplane(
+  p_1 %>%
+  layout(
+    scene = list(camera = list(eye = list(x = -0.3, y = 3, z = 1)))))
+
+p_4 <- add_vertical_hyperplane(p_4, threshold = 0.25)
+p_4 <- add_vertical_hyperplane(p_4, threshold = 0.75)
+
+@
+
+In logistic regression (binary case), we estimate the probability $\pix$. In 
+order to decide about the class of an observation, we set $\yh = 1$ iff 
+$\pixh \geq \alpha$ for some $\alpha \in \R$.
+
+\begin{enumerate}[a)]
+  \item Show that the decision boundary of the logistic classifier is a 
+  (linear!) hyperplane. \\
+  Hint: derive the value of $\thetab^T\xv$  (depending on 
+  $\alpha$) starting from which you predict $y = 1$ rather than $y = 0$.
+  \item Below you see the logistic function for a binary classification problem 
```

Comment:
Bei der Aufgabe verstehe ich noch nicht ganz, worauf Du rauswillst. Bei 1 und 4 ist theta_2 = 0 und theta_1 negativ? Und bei 2 und 3 alle negativ? Und alpha kann ich ja nicht wirklich ablesen (ja wobei, hinten hat man noch mal die Skala, vmtl. .25, .5 und .75?) - aber man sieht, dass der Wert von x2 egal ist in 4?

Reply:
Hast es schon erfasst eigentlich, ich will zeigen: bei 1 ist nur x_1 relevant, bei 2 & 3 beide features, allerdings ist theta für 3 größer (steilere softmax, höhere Konfidenz, mehr Werte nahe 0/1). Bei 4 gehts um alpha (genau, 0.25/0.5/0.75, wobei die exakten Werte egal sind): Je nach threshold genügen andere feauture values für Klasse 1, was dann mit unterschiedlichen Anforderungen an die ROC metrics zusammenhängt

Reply:
:+1: